### ISSUE TITLE
Crash - Update .. Returning statement on reference table

### ISSUE THREAD

**Issue by agedemenli** [citusdata/citus#6634]
Steps to reproduce:
```sql
create table t1 (a int);
select create_reference_table('t1');
update t1 set a = 1 returning
	(select pg_catalog.max(wal_bytes) from pg_catalog.pg_stat_wal)
         as c3;

```

**Comment by hanefi** [comment]
I found the reason of the crash in the function https://github.com/citusdata/citus/blob/93fcc5c5d85967812abcd41771d9b44900700019/src/backend/distributed/planner/combine_query_planner.c#L325-L326

If we add a check before casting, we can prevent the crashes.

```diff
             RangeTblFunction *rangeTblFunction = (RangeTblFunction *) linitial(
                 rangeTblEntry->functions);
+            if (!IsA(rangeTblFunction->funcexpr, FuncExpr))
+            {
+                return false;
+            }
             FuncExpr *funcExpr = castNode(FuncExpr, rangeTblFunction->funcexpr);
```

However, we end up with an invalid SQL command after rebuilding, and see the error message `input of anonymous composite types is not implemented` reported in a worker node.

```
DEBUG:  00000: query before rebuilding: UPDATE public.t1 SET a = 1 RETURNING (SELECT max(pg_stat_wal.wal_bytes) AS max FROM (SELECT w.wal_records, w.wal_fpi, w.wal_bytes, w.wal_buffers_full, w.wal_write, w.wal_sync, w.wal_write_time, w.wal_sync_time, w.stats_reset FROM pg_stat_get_wal() w(wal_records, wal_fpi, wal_bytes, wal_buffers_full, wal_write, wal_sync, wal_write_time, wal_sync_time, stats_reset)) pg_stat_wal) AS c3
LOCATION:  RebuildQueryStrings, deparse_shard_query.c:142
DEBUG:  00000: query after rebuilding:  UPDATE public.t1_102008 t1 SET a = 1 RETURNING (SELECT max(pg_stat_wal.wal_bytes) AS max FROM (SELECT w.wal_records, w.wal_fpi, w.wal_bytes, w.wal_buffers_full, w.wal_write, w.wal_sync, w.wal_write_time, w.wal_sync_time, w.stats_reset FROM CAST('(40,6,27364,0,12,12,0,0,"2023-01-24 13:19:16.027344+03")'::record AS record) w(wal_records, wal_fpi, wal_bytes, wal_buffers_full, wal_write, wal_sync, wal_write_time, wal_sync_time, stats_reset)) pg_stat_wal) AS c3
LOCATION:  RebuildQueryStrings, deparse_shard_query.c:153
```

In the rebuilt query, the CAST call is the root cause of the problem. We can not use an anonymous composite type as an input to `CAST()` in `CAST('(40,6,27364,0,12,12,0,0,"2023-01-24 13:19:16.027344+03")'::record AS record)`.

**Comment by hanefi** [comment]
Crash fixed by #6643 

See related error message improvement in issue #6651
